### PR TITLE
Forward-Compatibility with #591 (removed array like access)

### PR DESCRIPTION
**Pull request by mvorisek**
Only @romaninsh can review

### CODE REVIEW

**Comment by romaninsh** [comment]
@mvorisek, I think you need to re-create this again, too many conflicts.

As for array-access, can we implement like this (cc @DarkSide666):

``` php
foreach($user->ref('Invoices') as $row) {
  echo $row['ref'];
}

echo get_class($user->ref('Invoices'));  // Model
echo get_class($row);  // Entity
```

Additionally "Model::load(123)" would return entity. This makes it possible to refactor code in anticipation of dropping array access support and also implements a better pattern to break down Model into "set" and "entity"..

We could discuss this further, but Model->loaded() has been a thorn for a long time.

**Comment by mvorisek** [comment]
@romaninsh only for forward compatibility, otherwise already fully isolated and implemented in data as well ui - and bugs discovered due this :)

your example is fully supported, if you implement array access in all your models, you can continue using array like access as long as you want

(and there should be no single merge conflict like now on GH - I opened a ticket by GH for that to fix their code as well ;-) )

**Comment by romaninsh** [comment]
@mvorisek still seeing conflicts. 

Screenshot 2020-06-09 at 19 06 48

re-create or update this. anyways - i'm approving - if you create new PR i'll approve it too.

**Review by romaninsh — APPROVED** [review]
looks like we are on the way to drop array access, so thanks for this.

File: composer.json
Comment (mvorisek) [inline composer.json:32]: @romaninsh I think you should require this in all `legacy/2.0` branches in all repos, right?

**Comment by mvorisek** [comment]
@romaninsh after a few fixes, check and merge

**Pull request merged**
